### DISCUSSION TITLE
SINDYc custom library

### DISCUSSION THREAD

**Discussion by IamKFJ** [dynamicslab/pysindy#401] [General]
Hello everyone,
I am using a custom library and trying to use the feature where every variable has an associated library.
The documentation for this feature says that I should provide the library functions in the following shape:
`python (n_input_features, num_library_functions)`

What I am providing is the following (these are just basic functions I wanted to test this feature with. That's why they are mostly the same):

```python
 custom_functions = [[
                    lambda x, v, F, u: u,
                    lambda x: 1,
                  ],
                 [
                    lambda x, v, F, u: u,
                    lambda v: 1,
                 ],
                 [
                     lambda x, v, F, u: u,
                     lambda F: 1,
                 ],
                ]
``` 
My setup is the following:
- Three variables: x, v, F
- one control input: u
- optimizer: SR3

My code is the following
```python
opt_threshold = 0.013
dT = dT
    feature_names = ['x', 'v', 'F', 'u']

 custom_functions = [[
                    lambda x, v, F, u: u,
                    lambda x: 1,
                  ],
                 [
                    lambda x, v, F, u: u,
                    lambda v: 1,
                 ],
                 [
                     lambda x, v, F, u: u,
                     lambda F: 1,
                 ],
                ]
    lib_custom = ps.feature_library.CustomLibrary(library_functions=custom_functions)


    model = SINDy(feature_library=lib_custom, 
                  feature_names=feature_names,
                  optimizer=ps.SR3(normalize_columns=False, threshold=opt_threshold, thresholder='l1', max_iter=50)) 
    model.fit(x=X_Data, t=dT, u=U_Data, library_ensemble=True, ensemble=True, quiet=True)
``` 

While trying to fit the model always runs into the following Problem:
```
Traceback (most recent call last):
  File "filepath", line 68, in <module>
    X_sysident, method_characteristics, [t_SysIdent, t_generate] = sysident_create_data(X_train=X[:nData_SysIdent, :3], U_train=X[:nData_SysIdent, -1:], U_create=X[:, -1:], dT=X[0, 3],
  File "filepath", line 676, in sysident_create_data
    sysident_func, method_characteristics, t_SysIdent = SINDYc(X_Data=X_train, U_Data=U_train, dT=dT,
  File "filepath", line 563, in SINDYc
    model.fit(x=X_Data, t=dT, u=U_Data, library_ensemble=True, ensemble=True, quiet=True)
  File "filepath/python3.10/site-packages/pysindy/pysindy.py", line 414, in fit
    self.model.fit(x, x_dot)
  File "filepath/python3.10/site-packages/sklearn/base.py", line 1151, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "filepath/python3.10/site-packages/sklearn/pipeline.py", line 416, in fit
    Xt = self._fit(X, y, **fit_params_steps)
  File "filepath/python3.10/site-packages/sklearn/pipeline.py", line 370, in _fit
    X, fitted_transformer = fit_transform_one_cached(
  File "filepath/python3.10/site-packages/joblib/memory.py", line 353, in __call__
    return self.func(*args, **kwargs)
  File "filepath/python3.10/site-packages/sklearn/pipeline.py", line 950, in _fit_transform_one
    res = transformer.fit_transform(X, y, **fit_params)
  File "filepath/python3.10/site-packages/sklearn/utils/_set_output.py", line 140, in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
  File "filepath/python3.10/site-packages/sklearn/base.py", line 918, in fit_transform
    return self.fit(X, y, **fit_params).transform(X)
  File "filepath/python3.10/site-packages/pysindy/feature_library/base.py", line 191, in func
    result = wrapped_func(self, xs, *args, **kwargs)
  File "filepath/python3.10/site-packages/pysindy/feature_library/custom_library.py", line 176, in fit
    n_args = f.__code__.co_argcount
AttributeError: 'list' object has no attribute '__code__'. Did you mean: '__add__'?
``` 
I do not understand what I am am missing and where exactly my mistake lies.
Does anyone have experience with custom feature libraries ans supplying feature libraries for each variable individually?

How should the individual feature libraries (for x, v and F) be supplied to the custom feature library?
Are there any obvious mistakes I am making?

Id be very grateful for any tipps/help.
Thanks in advance!

**Comment by Jacob-Stevens-Haas** [comment]
you wrote
```python
 custom_functions = [[
                    lambda x, v, F, u: u,
                    lambda x: 1,
                  ],
                  ...
                ]
```

This is a list of list of functions.  You want a list of functions.  It would probably help if we included type annotations.

Secondly, you said:
> use the feature where every variable has an associated library

I think what you're looking for is the GeneralizedLibrary, which includes an argument `inputs_per_library`.  That will control which inputs (`x, v, F, u`) go to which input library.

As a third note, it looks like you're trying to add multiple constant functions.  This means you will have multiple identical functions, which can mess up the optimizer.

**Comment by odannis** [comment]
Hello,

Let say that I want to use SINDy on : 
$$\frac{dx}{dt} = \xi_1 f(x,y) + \xi_2 g(x,y)$$
$$\frac{dy}{dt} = \xi_3 h(x,y) + \xi_4 i(x,y)$$

I would rather not include functions $h$ and $i$ in the equation in the first equation $\frac{dx}{dt} = \xi_1 f(x,y) + \xi_2 g(x,y)$. How can I do that? 

It seems that whenever if I use the GeneralizedLibrary or the CustomLibrary, I would be in the case :
$$\frac{dx}{dt} = \xi_1 f(x,y) + \xi_2 g(x,y) + \xi_3 h(x,y) + \xi_4 i(x,y)$$
$$\frac{dy}{dt} = \xi_4 h(x,y) + \xi_5 i(x,y) +  \xi_6 f(x,y) + \xi_7 g(x,y)$$
And I would rather not add coefficient that I know they don't exist.

**Reply by Jacob-Stevens-Haas** [reply]
You have to use an optimizer that an apply constraints.  e.g. `ConstrainedSR3`.  Then you need to constrain the terms you don't want to have coefficients equal to zero

**Reply by odannis** [reply]
Thank you !

**Reply by Jacob-Stevens-Haas** [reply]
You're welcome!  Glad I could help.

**Comment by BK201-kkk** [comment]
I encountered the same problem. I also want to create an independent library for each ode. Have you solved this problem?

**Comment by Jacob-Stevens-Haas** [comment]
Ah, @BK201-kkk  pointed out the documentation that @IamKFJ was referencing.  It is indeed incorrect - you need a `GeneralizedLibrary`.  See #592 for details.
